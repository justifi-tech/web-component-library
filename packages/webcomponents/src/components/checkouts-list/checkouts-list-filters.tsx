import { Component, h, Prop } from '@stencil/core';
import { ICheckoutStatus } from '../../api';
import { filterParams, propsParams, clearParams } from './checkouts-list-params-state';
import { StyledHost } from '../../ui-components';
import { checkoutsListFilterMenu, paymentModeCheckoutsListFilterParam, statusCheckoutsListFilterParam } from '../../styles/parts';

@Component({
  tag: 'justifi-checkouts-list-filters',
  shadow: true
})
export class CheckoutsListFilters {
  @Prop() status?: ICheckoutStatus;
  @Prop() paymentMode?: string;

  componentWillLoad() {
    const propsToSet = {
      status: this.status,
      payment_mode: this.paymentMode,
    };

    Object.entries(propsToSet).forEach(([key, value]) => {
      if (value) {
        propsParams[key] = value;
      }
    });
  }
  
  setParamsOnChange = (name: string, value: string) => {
    filterParams[name] = value;
  }

  get checkoutStatusOptions(): { label: string, value: ICheckoutStatus | '' }[] {
    return [
      { label: 'All', value: '' },
      { label: 'Created', value: 'created' },
      { label: 'Completed', value: 'completed' },
      { label: 'Attempted', value: 'attempted' },
      { label: 'Expired', value: 'expired' },
    ]
  }

  get checkoutPaymentModeOptions(): { label: string, value: string }[] {
    return [
      { label: 'All', value: '' },
      { label: 'E-commerce', value: 'ecom' },
      { label: 'BNPL', value: 'bnpl' },
    ]
  }

  render() {
    const filterMenuParams = { ...filterParams }

    return (
      <StyledHost>
        <table-filters-menu params={filterMenuParams} clearParams={clearParams} part={checkoutsListFilterMenu}>
          <div class="grid-cols-2 gap-3 p-1">
            <div class="p-2">
              <form-control-select
                name="status"
                label="Status"
                options={this.checkoutStatusOptions}
                inputHandler={this.setParamsOnChange}
                defaultValue={this.status || filterParams.status}
                disabled={!!this.status}
                part={statusCheckoutsListFilterParam}
              />
            </div>
            <div class="p-2">
              <form-control-select
                name="payment_mode"
                label="Payment Mode"
                options={this.checkoutPaymentModeOptions}
                inputHandler={this.setParamsOnChange}
                defaultValue={this.paymentMode || filterParams.payment_mode}
                disabled={!!this.paymentMode}
                part={paymentModeCheckoutsListFilterParam}
              />
            </div>
          </div>
        </table-filters-menu>
      </StyledHost>
    )
  }
}
