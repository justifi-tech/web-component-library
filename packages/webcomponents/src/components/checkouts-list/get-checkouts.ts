import { Checkout } from '../../api';
import { ComponentErrorSeverity } from '../../api/ComponentError';
import { getErrorCode, getErrorMessage } from '../../api/services/utils';

export const makeGetCheckoutsList =
  ({ authToken, service, apiOrigin }) =>
    async ({ params, onSuccess, onError }) => {
      try {
        const response = await service.fetchCheckoutsList(authToken, params, apiOrigin);

        if (!response.error) {
          const pagingInfo = {
            ...response.page_info,
          };

          const checkouts =
            response.data?.map((dataItem) => new Checkout(dataItem)) || [];

          onSuccess({ checkouts, pagingInfo });
        } else {
          const responseError = getErrorMessage(response.error);
          const code = getErrorCode(response.error?.code);
          return onError({
            error: responseError,
            code,
            severity: ComponentErrorSeverity.ERROR,
          });
        }
      } catch (error) {
        const code = getErrorCode(error?.code);
        return onError({
          error: error.message || error,
          code,
          severity: ComponentErrorSeverity.ERROR,
        });
      }
    };
