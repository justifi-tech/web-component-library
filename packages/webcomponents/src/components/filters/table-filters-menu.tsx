import { Component, Listen, Prop, State, Watch, h } from '@stencil/core';
import { Button } from '../../ui-components/button';
import { hasFilters } from './utils';
import { dropdownMenu } from '../../styles/parts';

@Component({
  tag: 'table-filters-menu'
})
export class TableFiltersMenu {
  @State() anchorButton: HTMLElement;
  @State() menu: HTMLElement;
  @State() showClearButton: boolean = false;

  @Prop() params: any;
  @Prop() clearParams: () => void;

  @Watch('params')
  watchParamsHandler(newValue: any) {
    this.showClearButton = hasFilters(newValue);
  }

  @Listen('showEvent')
  show() {
    this.menu.classList.add('show');
  }

  @Listen('hideEvent')
  hide() {
    this.menu.classList.remove('show');
  }

  componentWillLoad() {
    this.showClearButton = hasFilters(this.params);
  }

  render() {
    return (
      <div class="d-flex gap-2">
        <div class="dropdown pb-3">
          <Button
            variant="primary"
            type="button dropdown-toggle"
            ref={(el) => (this.anchorButton = el)}
            class="btn btn-primary dropdown-toggle"
            data-test-id="open-filters-button"
            data-bs-toggle="dropdown"
          >
            {"Filters"}
          </Button>
          <custom-popper
            offset={[20, 10]}
            strategy="fixed"
            anchorRef={this.anchorButton}
            trigger="click"
          >
            <form
              class="dropdown-menu"
              part={dropdownMenu}
              data-test-id="filter-menu"
              ref={(el) => (this.menu = el)}
              onSubmit={(e) => e.preventDefault()}
            >
              {/* Render filter inputs here */}
              <slot />
            </form>
          </custom-popper>
        </div>
        <div>
          <Button
            variant="secondary"
            onClick={() => this.clearParams()}
            hidden={!this.showClearButton}
            data-test-id="clear-filters-button"
          >
            {"Clear Filters"}
          </Button>
        </div>
      </div>
    );
  }
}
