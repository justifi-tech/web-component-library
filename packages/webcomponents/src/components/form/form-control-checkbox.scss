@import 'reboot';
@import 'forms';

:host {
  @include host-base-styles;
  display: block;
}

// input[type='checkbox'][disabled] {
//   color: var(--jfi-form-control-disabled-color);
//   opacity: 0.5;
//   cursor: not-allowed;
// }

*::part(input[type='checkbox']) {
  margin: var(--jfi-form-control-margin);
  padding: var(--jfi-form-control-padding);
}