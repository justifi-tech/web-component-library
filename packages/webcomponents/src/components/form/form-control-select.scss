@import 'reboot';
@import 'forms';

:host {
  @include host-base-styles;
  display: block;
}

select[disabled] {
  background-color: #e9ecef;
  opacity: 0.5;
  cursor: not-allowed;
}
