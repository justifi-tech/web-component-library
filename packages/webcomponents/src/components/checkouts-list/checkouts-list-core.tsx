import { Component, h, Prop, State, Watch, Event, EventEmitter } from '@stencil/core';
import { Checkout, ICheckoutsParams, PagingInfo, SubAccount, pagingDefaults } from '../../api';
import { ComponentError } from '../../api/ComponentError';
import { tableExportedParts } from '../../ui-components/table/exported-parts';
import { onFilterChange } from '../../ui-components/filters/utils';
import { StyledHost, TableEmptyState, TableErrorState, TableLoadingState } from '../../ui-components';
import CheckoutsTable, { defaultColumnsKeys } from './checkouts-table';

@Component({
  tag: 'checkouts-list-core'
})
export class CheckoutsListCore {
  @Prop() getCheckouts: Function;
  @Prop() getSubAccounts: Function;
  @Prop() columns: string = defaultColumnsKeys;

  @State() checkouts: Checkout[] = [];
  @State() checkoutsTable: CheckoutsTable;
  @State() subAccounts: SubAccount[] = [];
  @State() loading: boolean = true;
  @State() errorMessage: string;
  @State() paging: PagingInfo = pagingDefaults;
  @State() params: ICheckoutsParams = {};

  @Watch('params')
  @Watch('getCheckouts')
  @Watch('getSubAccounts')
  @Watch('columns')
  updateOnPropChange() {
    this.fetchCheckouts();
  }

  @Event({
    eventName: 'checkout-row-clicked',
    bubbles: true,
  }) rowClicked: EventEmitter<Checkout>;

  @Event({ eventName: 'error-event' }) errorEvent: EventEmitter<ComponentError>;

  componentWillLoad() {
    if (this.getCheckouts && this.getSubAccounts) {
      this.fetchCheckouts();
    }
  }

  fetchCheckouts(): void {
    this.loading = true;

    this.getCheckouts({
      params: this.params,
      onSuccess: ({ checkouts, pagingInfo }) => {
        this.checkouts = checkouts;
        this.paging = pagingInfo;
        this.checkoutsTable = new CheckoutsTable(this.checkouts, this.columns);
        this.loading = false; // temp
      },
      onError: ({ error, code, severity }) => {
        this.errorMessage = error;
        this.errorEvent.emit({
          errorCode: code,
          message: error,
          severity,
        });
        this.loading = false;
      },
    });
  }

  async fetchSubAccounts(): Promise<void> {
    this.getSubAccounts({
      params: this.subAccountParams,
      onSuccess: ({ subAccounts }) => {
        this.subAccounts = subAccounts;
        this.checkouts = this.checkouts.map((checkout) => {
          checkout.sub_account_name = this.subAccounts.find((subAccount) => subAccount.id === checkout.account_id)?.name;
          return checkout;
        });
        this.loading = false;
      },
      onError: ({ error, code, severity }) => {
        this.errorMessage = error;
        this.errorEvent.emit({
          errorCode: code,
          message: error,
          severity,
        });
        this.loading = false;
      },
    });
  }

  handleClickPrevious = (beforeCursor: string) => {
    const newParams: any = { ...this.params };
    delete newParams.after_cursor;
    this.params = { ...newParams, before_cursor: beforeCursor };
  };

  handleClickNext = (afterCursor: string) => {
    const newParams: any = { ...this.params };
    delete newParams.before_cursor;
    this.params = { ...newParams, after_cursor: afterCursor };
  };

  rowClickHandler = (e) => {
    const clickedCheckoutID = e.target.closest('tr').dataset.rowEntityId;
    if (!clickedCheckoutID) return;
    this.rowClicked.emit(this.checkouts.find((checkout) => checkout.id === clickedCheckoutID));
  };

  setParamsOnChange = (name: string, value: string) => {
    let newParams = { [name]: value };
    this.params = onFilterChange(newParams, this.params);
  }

  clearParams = () => {
    this.errorMessage = '';
    this.params = {};
  }

  get subAccountParams() {
    let accountIdNumbers = this.checkouts.map((checkout) => checkout.account_id);
    let uniqueAccountIds = [...new Set(accountIdNumbers)];
    let accountIdString = uniqueAccountIds.join(',');
    return { sub_account_id: accountIdString };
  }

  get entityId() {
    return this.checkouts.map((checkout) => checkout.id);
  }

  get showEmptyState() {
    return !this.loading && !this.errorMessage && this.checkoutsTable.rowData.length < 1;
  }

  get showErrorState() {
    return !this.loading && !!this.errorMessage;
  }

  get showRowData() {
    return !this.showEmptyState && !this.showErrorState && !this.loading;
  }

  render() {
    return (
      <StyledHost exportparts={tableExportedParts}>
        <checkouts-list-filters
          params={this.params}
          setParamsOnChange={this.setParamsOnChange}
          clearParams={this.clearParams}
        />
        <div class="table-wrapper">
          <table class="table table-hover">
            <thead class="table-head sticky-top" part="table-head">
              <tr class="table-light text-nowrap" part="table-head-row">
                {this.checkoutsTable?.columnData?.map((column) => (
                  <th part="table-head-cell" scope="col" title={column.title}>
                    {column.label}
                  </th>
                ))}
              </tr>
            </thead>
            <tbody class="table-body" part="table-body">
              {/* fix this, not working with new checkoutsTable */}
              <TableLoadingState
                columnSpan={this.checkoutsTable?.columnData.length}
                isLoading={this.loading}
              />
              <TableEmptyState
                isEmpty={this.showEmptyState}
                columnSpan={this.checkoutsTable?.columnData.length}
              />
              <TableErrorState
                columnSpan={this.checkoutsTable?.columnData.length}
                errorMessage={this.errorMessage}
              />
              {this.showRowData && this.checkoutsTable?.rowData.map((data) => (<tr>{data}</tr>))}
            </tbody>
            {this.paging && (
              <tfoot class="sticky-bottom">
                <tr class="table-light align-middle">
                  <td part="pagination-bar" colSpan={this.checkoutsTable?.columnData?.length}>
                    <pagination-menu
                      paging={{
                        ...this.paging,
                        handleClickPrevious: this.handleClickPrevious,
                        handleClickNext: this.handleClickNext,
                      }}
                      params={this.params}
                    />
                  </td>
                </tr>
              </tfoot>
            )}
          </table>
        </div>
      </StyledHost>
    );
  }
}
