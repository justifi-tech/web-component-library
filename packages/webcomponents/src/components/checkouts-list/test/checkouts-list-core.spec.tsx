jest.mock('../../../ui-components/styled-host/styled-host.css', () => '');

import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { CheckoutsListCore } from '../checkouts-list-core';
import { PaginationMenu } from '../../pagination-menu/pagination-menu';
import mockSuccessResponse from '../../../../../../mockData/mockGetCheckoutsListSuccess.json';
import { IApiResponseCollection, ICheckout } from '../../../api';
import { makeGetCheckoutsList } from '../get-checkouts';
import { TableFiltersMenu } from '../../../ui-components/filters/table-filters-menu';
import { CheckoutsListFilters } from '../checkouts-list-filters';
import { SelectInput } from '../../../ui-components/form/form-control-select';

const mockCheckoutsListResponse = mockSuccessResponse as IApiResponseCollection<ICheckout[]>;
const components = [CheckoutsListCore, PaginationMenu, TableFiltersMenu, CheckoutsListFilters, SelectInput];

describe('checkouts-list-core', () => {
  it('renders properly with fetched data', async () => {
    const mockCheckoutsService = {
      fetchCheckoutsList: jest.fn().mockResolvedValue(mockCheckoutsListResponse),
    };

    const getCheckouts = makeGetCheckoutsList({
      accountId: 'mock_id',
      authToken: '123',
      service: mockCheckoutsService,
      apiOrigin: 'http://localhost:3000'
    });

    const page = await newSpecPage({
      components: [CheckoutsListCore, PaginationMenu],
      template: () => <checkouts-list-core getCheckoutsList={getCheckouts} />,
    });

    await page.waitForChanges();

    expect(page.rootInstance.checkouts[0]).toEqual(expect.objectContaining({ account_id: mockCheckoutsListResponse.data[0].account_id }));
    const rows = page.root.querySelectorAll('[data-test-id="table-row"]');
    expect(rows.length).toBe(mockCheckoutsListResponse.data.length);
    expect(mockCheckoutsService.fetchCheckoutsList).toHaveBeenCalled();
    expect(page.root).toMatchSnapshot();
  });

  it('displays an error state on failed data fetch', async () => {
    const mockService = {
      fetchCheckoutsList: jest.fn().mockRejectedValue(new Error('Fetch error'))
    };

    const getCheckouts = makeGetCheckoutsList({
      accountId: 'mock_id',
      authToken: 'some-auth-token',
      service: mockService,
      apiOrigin: 'http://localhost:3000'
    });

    const page = await newSpecPage({
      components: [CheckoutsListCore, PaginationMenu],
      template: () => <checkouts-list-core getCheckoutsList={getCheckouts} />,
    });

    await page.waitForChanges();

    expect(page.rootInstance.errorMessage).toBe('Fetch error');
    expect(page.root).toMatchSnapshot();
  });

  it('emits checkout-row-clicked event on row click', async () => {
    const mockCheckoutsService = {
      fetchCheckoutsList: jest.fn().mockResolvedValue(mockCheckoutsListResponse),
    };

    const getCheckouts = makeGetCheckoutsList({
      accountId: 'mock_id',
      authToken: '123',
      service: mockCheckoutsService,
      apiOrigin: 'http://localhost:3000'
    });

    const page = await newSpecPage({
      components: [CheckoutsListCore, PaginationMenu],
      template: () => <checkouts-list-core getCheckoutsList={getCheckouts} />,
    });

    await page.waitForChanges();

    const firstRow = page.root.querySelector('[data-test-id="table-row"]') as HTMLElement;
    expect(firstRow).not.toBeNull();

    const spyEvent = jest.fn();
    page.win.addEventListener('checkout-row-clicked', spyEvent);

    firstRow.click();
    expect(spyEvent).toHaveBeenCalled();
  });

  it('shows table filter menu on filter button click', async () => {
    const mockCheckoutsService = {
      fetchCheckoutsList: jest.fn().mockResolvedValue(mockCheckoutsListResponse),
    };

    const getCheckouts = makeGetCheckoutsList({
      accountId: 'mock_id',
      authToken: '123',
      service: mockCheckoutsService,
      apiOrigin: 'http://localhost:3000'
    });

    const page = await newSpecPage({
      components: [CheckoutsListCore, PaginationMenu, TableFiltersMenu, CheckoutsListFilters],
      template: () => <checkouts-list-core getCheckoutsList={getCheckouts} />,
    });

    await page.waitForChanges();

    const filterButton = page.root.querySelector('[data-test-id="open-filters-button"]') as HTMLElement;
    expect(filterButton).not.toBeNull();

    filterButton.click();
    await page.waitForChanges();

    const filterMenu = page.root.querySelector('[data-test-id="filter-menu"]') as HTMLElement;
    expect(filterMenu).not.toBeNull();
  });

  it('updates params and refetches data on filter interaction', async () => {
    const mockCheckoutsService = {
      fetchCheckoutsList: jest.fn().mockResolvedValue(mockCheckoutsListResponse),
    };

    const getCheckouts = makeGetCheckoutsList({
      accountId: 'mock_id',
      authToken: '123',
      service: mockCheckoutsService,
      apiOrigin: 'http://localhost:3000'
    });

    const page = await newSpecPage({
      components: components,
      template: () => <checkouts-list-core getCheckoutsList={getCheckouts} />,
    });

    const filterButton = page.root.querySelector('[data-test-id="open-filters-button"]') as HTMLElement;
    filterButton.click();

    const filterMenu = page.root.querySelector('checkouts-list-filters') as HTMLElement;
    expect(filterMenu).not.toBeNull();

    const selectFilter = filterMenu.querySelector('form-control-select') as HTMLFormControlSelectElement;
    expect(selectFilter).not.toBeNull();

    const selectFilterInput = selectFilter.querySelector('select') as HTMLSelectElement;
    expect(selectFilterInput).not.toBeNull();

    selectFilterInput.click();

    const selectOptions = selectFilterInput.querySelectorAll('option');
    expect(selectOptions).not.toBeNull();

    const succeededOption = selectOptions[3] as HTMLOptionElement;
    succeededOption.click();
    selectFilterInput.value = 'succeeded';
    selectFilterInput.dispatchEvent(new Event('input'));

    expect(mockCheckoutsService.fetchCheckoutsList).toHaveBeenCalledTimes(2);
    const updatedParams = page.rootInstance.params;
    expect(updatedParams).toEqual({ "status": "succeeded" });
  });

  it('clears filters and refetches data on clear filters interaction', async () => {
    const mockCheckoutsService = {
      fetchCheckoutsList: jest.fn().mockResolvedValue(mockCheckoutsListResponse),
    };

    const getCheckouts = makeGetCheckoutsList({
      accountId: 'mock_id',
      authToken: '123',
      service: mockCheckoutsService,
      apiOrigin: 'http://localhost:3000'
    });

    const page = await newSpecPage({
      components: components,
      template: () => <checkouts-list-core getCheckoutsList={getCheckouts} />,
    });

    const filterButton = page.root.querySelector('[data-test-id="open-filters-button"]') as HTMLElement;
    filterButton.click();

    const filterMenu = page.root.querySelector('table-filters-menu') as HTMLElement;
    expect(filterMenu).not.toBeNull();

    const selectFilter = filterMenu.querySelector('form-control-select') as HTMLFormControlSelectElement;
    expect(selectFilter).not.toBeNull();

    const selectFilterInput = selectFilter.querySelector('select') as HTMLSelectElement;
    expect(selectFilterInput).not.toBeNull();

    selectFilterInput.click();

    const selectOptions = selectFilterInput.querySelectorAll('option');
    expect(selectOptions).not.toBeNull();

    const succeededOption = selectOptions[3] as HTMLOptionElement;
    succeededOption.click();
    selectFilterInput.value = 'succeeded';
    selectFilterInput.dispatchEvent(new Event('input'));

    const clearButton = filterMenu.querySelector('[data-test-id="clear-filters-button"]') as HTMLElement;
    expect(clearButton).not.toBeNull();

    clearButton.click();

    expect(mockCheckoutsService.fetchCheckoutsList).toHaveBeenCalledTimes(3);
    const updatedParams = page.rootInstance.params;
    expect(updatedParams).toEqual({});
  });

  it('updates params and refetches data on pagination interaction', async () => {
    const mockCheckoutsService = {
      fetchCheckoutsList: jest.fn().mockResolvedValue(mockCheckoutsListResponse),
    };

    const getCheckouts = makeGetCheckoutsList({
      accountId: 'mock_id',
      authToken: '123',
      service: mockCheckoutsService,
      apiOrigin: 'http://localhost:3000'
    });

    const page = await newSpecPage({
      components: [CheckoutsListCore, PaginationMenu],
      template: () => <checkouts-list-core getCheckoutsList={getCheckouts} />,
    });

    await page.waitForChanges();

    // Assuming handleClickNext is accessible and modifies `params` which triggers a re-fetch
    page.rootInstance.handleClickNext('nextCursor');
    await page.waitForChanges();

    // The mock function should be called 2 times: once for the initial load and later after the pagination interaction
    expect(mockCheckoutsService.fetchCheckoutsList).toHaveBeenCalledTimes(2);
    const updatedParams = page.rootInstance.params;
    expect(updatedParams.after_cursor).toBe('nextCursor');
  });

  it('emits error event on fetch error', async () => {
    const mockService = {
      fetchCheckoutsList: jest.fn().mockRejectedValue(new Error('Fetch error'))
    };

    const getCheckouts = makeGetCheckoutsList({
      accountId: 'mock_id',
      authToken: 'some-auth',
      service: mockService,
      apiOrigin: 'http://localhost:3000'
    });

    const errorEvent = jest.fn();

    const page = await newSpecPage({
      components: [CheckoutsListCore, PaginationMenu],
      template: () => <checkouts-list-core getCheckoutsList={getCheckouts} onError-event={errorEvent} />,
    });

    await page.waitForChanges();

    expect(errorEvent).toHaveBeenCalledWith(
      expect.objectContaining({
        detail: {
          errorCode: 'fetch-error',
          message: 'Fetch error',
          severity: 'error',
        }
      })
    );
  });
});
