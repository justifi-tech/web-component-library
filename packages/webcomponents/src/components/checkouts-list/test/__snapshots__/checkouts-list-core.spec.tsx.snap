// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkouts-list-core displays an error state on failed data fetch 1`] = `
<checkouts-list-core>
  <div>
    <checkouts-list-filters>
      <table-filters-menu>
        <!---->
        <div class="d-flex gap-2">
          <div class="dropdown pb-3">
            <button class="btn btn-primary dropdown-toggle" data-bs-toggle="dropdown" data-test-id="open-filters-button" part="button button-primary" type="button dropdown-toggle" variant="primary">
              Filters
            </button>
            <custom-popper strategy="fixed" trigger="click">
              <form class="dropdown-menu" data-test-id="filter-menu" part="dropdown-menu">
                <div class="gap-3 grid-cols-2 p-1">
                  <div class="p-2">
                    <form-control-select exportparts="label,input,input-invalid">
                      <div class="d-flex flex-column form-group">
                        <div class="align-items-start d-flex gap-2">
                          <label class="form-label" htmlfor="status" part="label">
                            Status
                          </label>
                          <form-control-tooltip></form-control-tooltip>
                        </div>
                        <select class="form-select" id="status" name="status" part="input ">
                          <option value="">
                            All
                          </option>
                          <option value="created">
                            Created
                          </option>
                          <option value="completed">
                            Completed
                          </option>
                          <option value="attempted">
                            Attempted
                          </option>
                          <option value="expired">
                            Expired
                          </option>
                        </select>
                        <form-control-error-text name="status"></form-control-error-text>
                      </div>
                    </form-control-select>
                  </div>
                  <div class="p-2">
                    <form-control-select exportparts="label,input,input-invalid">
                      <div class="d-flex flex-column form-group">
                        <div class="align-items-start d-flex gap-2">
                          <label class="form-label" htmlfor="payment_mode" part="label">
                            Payment Mode
                          </label>
                          <form-control-tooltip></form-control-tooltip>
                        </div>
                        <select class="form-select" id="payment_mode" name="payment_mode" part="input ">
                          <option value="">
                            All
                          </option>
                          <option value="ecom">
                            E-commerce
                          </option>
                          <option value="bnpl">
                            BNPL
                          </option>
                        </select>
                        <form-control-error-text name="payment_mode"></form-control-error-text>
                      </div>
                    </form-control-select>
                  </div>
                </div>
              </form>
            </custom-popper>
          </div>
          <div>
            <button class="btn btn-secondary" data-test-id="clear-filters-button" hidden="" part="button button-secondary" variant="secondary">
              Clear Filters
            </button>
          </div>
        </div>
      </table-filters-menu>
    </checkouts-list-filters>
    <div class="table-wrapper">
      <table class="table table-hover">
        <thead class="sticky-top table-head" part="table-head">
          <tr class="table-light text-nowrap" part="table-head-row">
            <th part="table-head-cell" scope="col" title="The date each checkout occurred">
              Processed On
            </th>
            <th part="table-head-cell" scope="col" title="The dollar amount of each checkout">
              Payment Amount
            </th>
            <th part="table-head-cell" scope="col" title="The description of each checkout">
              Payment Description
            </th>
            <th part="table-head-cell" scope="col" title="The payment mode of each checkout">
              Payment Mode
            </th>
            <th part="table-head-cell" scope="col" title="The current status of each checkout">
              Status
            </th>
          </tr>
        </thead>
        <tbody class="table-body" part="table-body">
          <tr>
            <td class="error-state" colspan="5" data-test-id="table-error-state" part="error-state" style="text-align: center;">
              An unexpected error occurred: Fetch error
            </td>
          </tr>
        </tbody>
        <tfoot class="sticky-bottom">
          <tr class="align-middle table-light">
            <td colspan="5" part="pagination-bar">
              <pagination-menu exportparts="page-arrow,page-button,page-button-disabled,page-button-text">
                <nav aria-label="Table pagination" class="d-flex gap-3 justify-content-end">
                  <ul class="m-0 pagination">
                    <li class="disabled page-item">
                      <a class="page-link" href="#" part="page-button page-button-disabled">
                        <span class="me-1" part="page-arrow">
                          «
                        </span>
                        <span part="page-button-text">
                          Previous
                        </span>
                      </a>
                    </li>
                    <li class="disabled page-item">
                      <a class="page-link" href="#" part="page-button page-button-disabled">
                        <span part="page-button-text">
                          Next
                        </span>
                        <span class="ms-1" part="page-arrow">
                          »
                        </span>
                      </a>
                    </li>
                  </ul>
                </nav>
              </pagination-menu>
            </td>
          </tr>
        </tfoot>
      </table>
    </div>
  </div>
</checkouts-list-core>
`;

exports[`checkouts-list-core renders properly with fetched data 1`] = `
<checkouts-list-core>
  <div>
    <checkouts-list-filters>
      <table-filters-menu>
        <!---->
        <div class="d-flex gap-2">
          <div class="dropdown pb-3">
            <button class="btn btn-primary dropdown-toggle" data-bs-toggle="dropdown" data-test-id="open-filters-button" part="button button-primary" type="button dropdown-toggle" variant="primary">
              Filters
            </button>
            <custom-popper strategy="fixed" trigger="click">
              <form class="dropdown-menu" data-test-id="filter-menu" part="dropdown-menu">
                <div class="gap-3 grid-cols-2 p-1">
                  <div class="p-2">
                    <form-control-select exportparts="label,input,input-invalid">
                      <div class="d-flex flex-column form-group">
                        <div class="align-items-start d-flex gap-2">
                          <label class="form-label" htmlfor="status" part="label">
                            Status
                          </label>
                          <form-control-tooltip></form-control-tooltip>
                        </div>
                        <select class="form-select" id="status" name="status" part="input ">
                          <option value="">
                            All
                          </option>
                          <option value="created">
                            Created
                          </option>
                          <option value="completed">
                            Completed
                          </option>
                          <option value="attempted">
                            Attempted
                          </option>
                          <option value="expired">
                            Expired
                          </option>
                        </select>
                        <form-control-error-text name="status"></form-control-error-text>
                      </div>
                    </form-control-select>
                  </div>
                  <div class="p-2">
                    <form-control-select exportparts="label,input,input-invalid">
                      <div class="d-flex flex-column form-group">
                        <div class="align-items-start d-flex gap-2">
                          <label class="form-label" htmlfor="payment_mode" part="label">
                            Payment Mode
                          </label>
                          <form-control-tooltip></form-control-tooltip>
                        </div>
                        <select class="form-select" id="payment_mode" name="payment_mode" part="input ">
                          <option value="">
                            All
                          </option>
                          <option value="ecom">
                            E-commerce
                          </option>
                          <option value="bnpl">
                            BNPL
                          </option>
                        </select>
                        <form-control-error-text name="payment_mode"></form-control-error-text>
                      </div>
                    </form-control-select>
                  </div>
                </div>
              </form>
            </custom-popper>
          </div>
          <div>
            <button class="btn btn-secondary" data-test-id="clear-filters-button" hidden="" part="button button-secondary" variant="secondary">
              Clear Filters
            </button>
          </div>
        </div>
      </table-filters-menu>
    </checkouts-list-filters>
    <div class="table-wrapper">
      <table class="table table-hover">
        <thead class="sticky-top table-head" part="table-head">
          <tr class="table-light text-nowrap" part="table-head-row">
            <th part="table-head-cell" scope="col" title="The date each checkout occurred">
              Processed On
            </th>
            <th part="table-head-cell" scope="col" title="The dollar amount of each checkout">
              Payment Amount
            </th>
            <th part="table-head-cell" scope="col" title="The description of each checkout">
              Payment Description
            </th>
            <th part="table-head-cell" scope="col" title="The payment mode of each checkout">
              Payment Mode
            </th>
            <th part="table-head-cell" scope="col" title="The current status of each checkout">
              Status
            </th>
          </tr>
        </thead>
        <tbody class="table-body" part="table-body">
          <tr data-row-entity-id="cho_1a2b3c4d5e6f7g8h9i0j" data-test-id="table-row" part="table-row table-row-odd">
            <td part="table-cell">
              <div class="fw-bold">
                Aug 8, 2024
              </div>
              <div class="fw-bold">
                3:47pm
              </div>
            </td>
            <td part="table-cell">
              $3.99 USD
            </td>
            <td part="table-cell">
              One Chocolate Donut
            </td>
            <td part="table-cell">
              Card Present
            </td>
            <td part="table-cell">
              <span class="badge bg-success" title="The checkout has been completed">
                Completed
              </span>
            </td>
          </tr>
          <tr data-row-entity-id="cho_2b3c4d5e6f7g8h9i0j1a" data-test-id="table-row" part="table-row table-row-even">
            <td part="table-cell">
              <div class="fw-bold">
                Aug 8, 2024
              </div>
              <div class="fw-bold">
                4:32pm
              </div>
            </td>
            <td part="table-cell">
              $10.99 USD
            </td>
            <td part="table-cell">
              1 Dozen Chocolate Donuts
            </td>
            <td part="table-cell">
              Card Present
            </td>
            <td part="table-cell">
              <span class="badge bg-danger" title="The checkout has expired">
                Expired
              </span>
            </td>
          </tr>
          <tr data-row-entity-id="cho_3c4d5e6f7g8h9i0j1a2b" data-test-id="table-row" part="table-row table-row-odd">
            <td part="table-cell">
              <div class="fw-bold">
                Aug 8, 2024
              </div>
              <div class="fw-bold">
                4:32pm
              </div>
            </td>
            <td part="table-cell">
              $19.99 USD
            </td>
            <td part="table-cell">
              2 Dozen Chocolate Donuts - Website To Go Order
            </td>
            <td part="table-cell">
              E-commerce
            </td>
            <td part="table-cell">
              <span class="badge bg-primary" title="The checkout has been created">
                Created
              </span>
            </td>
          </tr>
          <tr data-row-entity-id="cho_4d5e6f7g8h9i0j1a2b3c" data-test-id="table-row" part="table-row table-row-even">
            <td part="table-cell">
              <div class="fw-bold">
                Aug 8, 2024
              </div>
              <div class="fw-bold">
                4:32pm
              </div>
            </td>
            <td part="table-cell">
              $89.99 USD
            </td>
            <td part="table-cell">
              Weekly Donut Box Subscription
            </td>
            <td part="table-cell">
              Buy Now Pay Later
            </td>
            <td part="table-cell">
              <span class="badge bg-primary" title="The checkout has been created">
                Created
              </span>
            </td>
          </tr>
          <tr data-row-entity-id="cho_5e6f7g8h9i0j1a2b3c4d" data-test-id="table-row" part="table-row table-row-odd">
            <td part="table-cell">
              <div class="fw-bold">
                Aug 8, 2024
              </div>
              <div class="fw-bold">
                4:32pm
              </div>
            </td>
            <td part="table-cell">
              $24.99 USD
            </td>
            <td part="table-cell">
              Donut Shop Size M T Shirt - Website Order
            </td>
            <td part="table-cell">
              E-commerce
            </td>
            <td part="table-cell">
              <span class="badge bg-secondary" title="The checkout has been attempted">
                Attempted
              </span>
            </td>
          </tr>
        </tbody>
        <tfoot class="sticky-bottom">
          <tr class="align-middle table-light">
            <td colspan="5" part="pagination-bar">
              <pagination-menu exportparts="page-arrow,page-button,page-button-disabled,page-button-text">
                <nav aria-label="Table pagination" class="d-flex gap-3 justify-content-end">
                  <ul class="m-0 pagination">
                    <li class="disabled page-item">
                      <a class="page-link" href="#" part="page-button page-button-disabled">
                        <span class="me-1" part="page-arrow">
                          «
                        </span>
                        <span part="page-button-text">
                          Previous
                        </span>
                      </a>
                    </li>
                    <li class="disabled page-item">
                      <a class="page-link" href="#" part="page-button page-button-disabled">
                        <span part="page-button-text">
                          Next
                        </span>
                        <span class="ms-1" part="page-arrow">
                          »
                        </span>
                      </a>
                    </li>
                  </ul>
                </nav>
              </pagination-menu>
            </td>
          </tr>
        </tfoot>
      </table>
    </div>
  </div>
</checkouts-list-core>
`;
